synapse/tools/migrate_020.py

```diff
@@ -0,0 +1,1652 @@
+
+            yield {
+                'migrop': splt[0].decode(),
+                'logtyp': splt[1].decode(),
+                'key': skey,
+                'val': s_msgpack.un(lval),
+            }
+
+    #############################################################
+    # Source (0.1.x) operations
+    #############################################################
+
+    async def _srcPackNode(self, buid, ndef, props, tags, tagprops):
+        '''
+        Return a packaged node
+        '''
+        return (buid, {
+            'ndef': ndef,
+            'props': props,
+            'tags': tags,
+            'tagprops': tagprops,
+        })
+
+    async def _srcIterForms(self, buidslab, buiddb):
+        '''
+        Iterate only to retrieve literal node forms.
+
+        Yields:
+            (str): Form name for a unique node
+        '''
+        for lkey, _ in buidslab.scanByFull(db=buiddb):
+            prop = lkey[32:].decode('utf8')
+            if prop[0] == '*':
+                yield prop[1:]
+
+    async def _srcIterNodes(self, buidslab, buiddb):
+        '''
+        Yield node information directly from the 0.1.x source slab.
+
+        Yields:
+            (tuple):
+                (<buid>, {
+                    'ndef': (<formname>, <formvalu>),
+                    'props': {<propname>: <propvalu>, ...},
+                    'tags': {<tagname>: <tagvalu>, ...},
+                    'tagprops': {
+                        <tagname>: {<propname>: <propvalu>, ...},
+                        ...
+                    }
+                )
+        '''
+        buid = None
+        ndef = None
+        props = {}
+        tags = {}
+        tagprops = collections.defaultdict(dict)
+        for lkey, lval in buidslab.scanByFull(db=buiddb):
+            rowbuid = lkey[0:32]
+            prop = lkey[32:].decode('utf8')
+            valu, indx = s_msgpack.un(lval)  # throwing away indx
+
+            # new node; if not at start, yield the last node and reset
+            if buid is not None and rowbuid != buid:
+                yield await self._srcPackNode(buid, ndef, props, tags, tagprops)
+                buid = None
+                ndef = None
+                props = {}
+                tags = {}
+                tagprops = collections.defaultdict(dict)
+
+            if buid is None:
+                buid = rowbuid
+
+            # add node information
+            if prop[0] == '*':
+                if ndef is None:
+                    ndef = (prop, valu)
+                else:
+                    props[prop] = valu
+
+            elif prop[0] == '#':
+                if ':' in prop:  # tagprop
+                    tname, tprop = prop.split(':')
+                    tagprops[tname][tprop] = valu
+                else:
+                    tags[prop] = valu
+
+            else:
+                props[prop] = valu
+
+        # yield last node
+        yield await self._srcPackNode(buid, ndef, props, tags, tagprops)
```

Comment:
If for some reason a cortex with no nodes is being migrated, this will yield an invalid node.